File: apps/windows/src/Shared/TapStateMachine.cs
using System;

namespace VoiceVector.Shared
{
    /// <summary>
    /// Pure hotkey-gesture logic — behavior-identical to the macOS app (see
    /// docs/config-schema.md "Hotkey semantics"). Recording starts on the very
    /// first key-down; a stray single tap is discarded after the tap window.
    /// </summary>
    public class TapStateMachine
    {
        public enum Act { StartRecording, Commit, Discard }

        public enum PhaseKind { Idle, Pressed, AwaitingSecondTap, Latched, Draining }

        public TapStartMode StartMode { get; private set; }
        public double HoldThreshold = 0.35;
        public double TapWindow = 0.40;

        public PhaseKind Phase { get; private set; }
        private double _pressedSince;
        private int _tap;
        private double _deadline;

        public TapStateMachine(TapStartMode startMode)
        {
            StartMode = startMode;
            Phase = PhaseKind.Idle;
        }

        public double? PendingDeadline
        {
            get { return Phase == PhaseKind.AwaitingSecondTap ? _deadline : (double?)null; }
        }

        public bool IsActive { get { return Phase != PhaseKind.Idle; } }

        private int TapsRequired { get { return StartMode == TapStartMode.DoubleTap ? 2 : 1; } }

        public Act[] KeyDown(double now)
        {
            switch (Phase)
            {
                case PhaseKind.Idle:
                    Phase = PhaseKind.Pressed;
                    _pressedSince = now;
                    _tap = 1;
                    return new[] { Act.StartRecording };
                case PhaseKind.AwaitingSecondTap:
                    Phase = PhaseKind.Pressed;
                    _pressedSince = now;
                    _tap = 2;
                    return Array.Empty<Act>();
                case PhaseKind.Latched:
                    Phase = PhaseKind.Draining;
                    return new[] { Act.Commit };
                default:
                    return Array.Empty<Act>();
            }
        }

        public Act[] KeyUp(double now)
        {
            switch (Phase)
            {
                case PhaseKind.Pressed:
                    if (now - _pressedSince >= HoldThreshold)
                    {
                        Phase = PhaseKind.Idle;
                        return new[] { Act.Commit }; // hold-to-talk
                    }
                    if (_tap >= TapsRequired)
                    {
                        Phase = PhaseKind.Latched;
                        return Array.Empty<Act>();
                    }
                    Phase = PhaseKind.AwaitingSecondTap;
                    _deadline = now + TapWindow;
                    return Array.Empty<Act>();
                case PhaseKind.Draining:
                    Phase = PhaseKind.Idle;
                    return Array.Empty<Act>();
                default:
                    return Array.Empty<Act>();
            }
        }

        public Act[] Expire(double now)
        {
            if (Phase == PhaseKind.AwaitingSecondTap && now >= _deadline)
            {
                Phase = PhaseKind.Idle;
                return new[] { Act.Discard };
            }
            return Array.Empty<Act>();
        }

        public Act[] Cancel()
        {
            bool wasActive = IsActive;
            Phase = PhaseKind.Idle;
            return wasActive ? new[] { Act.Discard } : Array.Empty<Act>();
        }
    }
}

File: apps/macos/Sources/VoiceVector/Hotkey/TapStateMachine.swift
import Foundation

/// Pure logic for interpreting hotkey presses. Recording starts on the very
/// first key-down (so no speech is clipped); a stray single tap is discarded
/// after the tap window expires.
///
/// Semantics (both always active):
/// - Hold-to-talk: press, speak, release (≥ holdThreshold) → commit.
/// - Toggle: double-tap (or single tap, per config) starts; next tap commits.
struct TapStateMachine {
    enum Action: Equatable {
        case startRecording
        case commit
        case discard
    }

    enum Phase: Equatable {
        case idle
        /// Key is down; `tap` is 1 for the first press, 2 for the second.
        case pressed(since: TimeInterval, tap: Int)
        /// First short tap released; waiting to see if a second tap arrives.
        case awaitingSecondTap(deadline: TimeInterval)
        /// Toggle recording in progress, key up.
        case latched
        /// Commit issued on key-down; swallowing the matching key-up.
        case draining
    }

    var startMode: TapStartMode
    var holdThreshold: TimeInterval = 0.35
    var tapWindow: TimeInterval = 0.40

    private(set) var phase: Phase = .idle

    /// Deadline (absolute time) at which `expire(now:)` must be called, if any.
    var pendingDeadline: TimeInterval? {
        if case .awaitingSecondTap(let deadline) = phase { return deadline }
        return nil
    }

    var isActive: Bool { phase != .idle }

    mutating func keyDown(at now: TimeInterval) -> [Action] {
        switch phase {
        case .idle:
            phase = .pressed(since: now, tap: 1)
            return [.startRecording]
        case .awaitingSecondTap:
            phase = .pressed(since: now, tap: 2)
            return []
        case .latched:
            phase = .draining
            return [.commit]
        case .pressed, .draining:
            return [] // auto-repeat or duplicate events
        }
    }

    mutating func keyUp(at now: TimeInterval) -> [Action] {
        switch phase {
        case .pressed(let since, let tap):
            if now - since >= holdThreshold {
                // Held: push-to-talk (works on the first or second press).
                phase = .idle
                return [.commit]
            }
            if tap >= tapsRequired {
                phase = .latched
                return []
            }
            phase = .awaitingSecondTap(deadline: now + tapWindow)
            return []
        case .draining:
            phase = .idle
            return []
        case .idle, .awaitingSecondTap, .latched:
            return []
        }
    }

    /// Call when `pendingDeadline` passes without a second tap.
    mutating func expire(at now: TimeInterval) -> [Action] {
        if case .awaitingSecondTap(let deadline) = phase, now >= deadline {
            phase = .idle
            return [.discard]
        }
        return []
    }

    /// External cancel (Esc) or forced reset.
    mutating func cancel() -> [Action] {
        let wasActive = isActive
        phase = .idle
        return wasActive ? [.discard] : []
    }

    private var tapsRequired: Int { startMode == .doubleTap ? 2 : 1 }
}
